### PR TITLE
Avoid giving 'no best type' error when one of the types contains error types

### LINKED ISSUE #67751: Don't produce a "best type not found" error because of an ErrorType in a nested position
**Version Used**: 17.6 Preview 3

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzggPgAQMwAJMBMuAwrgN4CwAULrfnpgCy4CyADABQCU51d/rAIwcYAUwDuAbQC65XAFEASooDyiuLkG4AvlwDcuAPSHcKAI4BXKAHNrAG1G4A9jDsBPZzFwByJasXefAK0LAQiEjJyHH5qGoJcGhyCGgQ8ugbGppY29o4uuOIAFk4OPmJS0t64ogAeAA4ATqIoKFAuAHRBtNpduL04+MwswrAIkRANDTxkOv0MQ2EcoxqjXOOT07NU2kA=)

```cs
public class C {
    public void M0() {
        M1(new[] { ERROR, 1 }); // squiggle only on 'ERROR'
        M2(new[] { (ERROR, 1), (1, 2) }); // squiggle on whole 'new[]' expression.
    }
    
    public void M1(int[] arr) { }
    public void M2((int, int)[] arr) { }
}
```

**Expected behavior**: squiggle only on 'ERROR' expression in both calls.
**Actual behavior**: squiggle on whole 'new[]' expression in call to 'M2'.

This confused me/slowed me down so want to grab an hour and fix this one up really quick.

### PR DESCRIPTION
**Pull request by RikkiGibson**
Closes #67751

Moving to draft while I build out the test coverage.

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs [hunk 1]
```diff
@@ -24669,5 +24669,149 @@ public void ArrayGenericConstraintTypes()
...
+
+        [Fact]
+        public void BestType_NestedError_LambdaReturns()
+        {
+            var source = """
+                using System;
+
+                public class C {
+                    public void M0() {
+                        Func<bool, int> fn1 = flag =>
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs:24787]: Could you also include a lambda scenario without target-type (ie. `var fn = ...`)?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -57993,18 +57993,18 @@ static void G(C? x2, Unknown y2)
...
+                // 0.cs(3,25): error CS0246: The type or namespace name 'Unknown' could not be found (are you missing a using directive or an assembly reference?)
                 //     static void F(C x1, Unknown? y1)
                 Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "Unknown").WithArguments("Unknown").WithLocation(3, 25),
-                // (5,18): error CS0826: No best type found for implicitly-typed array
+                // 0.cs(5,26): error CS0029: Cannot implicitly convert type 'C' to 'Unknown?'
                 //         var a1 = new[] { x1, y1 };
-                Diagnostic(ErrorCode.ERR_ImplicitlyTypedArrayNoBestType, "new[] { x1, y1 }").WithLocation(5, 18),
-                // (7,18): error CS0826: No best type found for implicitly-typed array
+                Diagnostic(ErrorCode.ERR_NoImplicitConv, "x1").WithArguments("C", "Unknown?").WithLocation(5, 26),
+                // 0.cs(7,30): error CS0029: Cannot implicitly convert type 'C' to 'Unknown?'
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:58005]: Not blocking this PR: should we also prevent this cascading diagnostic?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2) with a question and a test suggestion to consider

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:58005]: I think that's a good idea. Opening #69138.

**Pull request merged**